### ISSUE TITLE
#[extendr] macro does not accept `Self` as an argument in metho impls

### ISSUE THREAD

**Issue by JosiahParry** [extendr/extendr#547]
When implementing methods for a structs that are made available to R via extendr, the methods cannot contain `Self`. The compiler erros with the following: 

```
Self` is only available in impls, traits, and type definitions
  |        |
  |        `Self` not allowed in a function
  ```

contents of lib.rs are below

```r lib.rs
use extendr_api::prelude::*;

// define a struct
struct VecUsize(pub Vec<Option<usize>>);

// if remove this comment extendr will compile
#[extendr]
impl VecUsize {
    fn extend(&self, y: Self) -> &Self {
        self
    }
}

extendr_module! {
    mod buggg;    
}
```

To reproduce, create include the above struct and impl in an extendr project.

**Comment by multimeric** [comment]
Yep, I think we could resolve `Self` to the struct name ourselves here, this should be doable. I think the workaround might be to make the return type be the actual Struct type.

**Comment by JosiahParry** [comment]
I tried this but it didn't work. I think it's an infinite recursion type of issue. Because if I want to return `VecUsize` in the impl but I haven't yet completed the struct impl it doesn't know how to return it
